### ISSUE TITLE
feature : implement a BEMIT compatible customisable syntax checker

### ISSUE THREAD

**Issue by gael-boyenval** [stylelint/stylelint#2698]
## Feature : a customizable BEMIT selector syntax checker 
the BEM linter is great but it doesn't allow to customize enough to enforce BEMIT @csswizardry (Harry Roberts) proposal.

[a great article about BEMIT here](https://csswizardry.com/2015/08/bemit-taking-the-bem-naming-convention-a-step-further/) or for the lazy ones : [ BEMIT Cheatsheet ](https://gist.github.com/stephenway/a6145d9b4430e8c55a77)

it's about namespaced BEM selectors, and enforce different rules for those deferents types.

some of the BEMIT types:
- scope `.s-`
- themes `.t-`
- components `.c-`
- states `.is- | .has- | .no- ` ...

some "types" can be combined : 
`.c-component.is-open`  => ok

but others not : 
`.c-component-a .c-component-b`  => warn

implementing a strict ruleset would be great but being able to configure it and create our own would be even better : 

```js
{
    // basic syntax checker, create your own flavor of BEM if you want
    generalSyntax: {
        wordDelimiterStyle: 'lowerCamelCase' | 'CamelCase'  | 'snake_case' | 'kebab-case' | 'min' | 'CAPS',
        modifierDelimiter: '__' | regex,
        elementDelimiter: '__' | regex
    },
    registerTypes: [
        {
            name: 'components', // the name of the type
            namespaces: .c-|.custom-|.other-|, // a regex to match namespaces,
            canCallClassType: ['states'] | 'any', // array of registred types that can be called in a selector`.c-popup.is-open`
            allowHostDiferentBlock: false // disalow `.c-carousel__nav .c-btn`
        },{
            name: 'states',
            namespaces: .is-|.has-|.no-|,
            canCallClassType: 'any',
        },{
            name: 'utilities',
            namespaces: /.is-|.has-|.no-|/ regex,
            canCallClassType: ['registredName'] | 'any' | 'none',
        },{
            name: 'themes',
            namespaces: /.t-/ regex,
            canCallClassType: 'any', // allow '.t-dark .c-btn--alert' selector
        },{
            name: 'scopes',
            namespaces: /.t-/ regex,
            canCallClassType: 'any',
        }
        ...
    ]
}
```

each selectors would be evaluated, the first class would be considered has the "host" and all other has "hosted" :

#### exemple :  

with the given registered types : 

```js
      {
            name: 'components',
            namespaces: 'c-',
            canCallClassType: ['states'],
            allowHostDiferentBlock: false
        },{
            name: 'states',
            namespaces: 'is-|has-|no-',
            canCallClassType: any,
            allowHostDiferentBlock: true
        },{
            name: 'hacks',
            namespaces: '_',
            canCallClassType: any,
            allowHostDiferentBlock: true
        }
...
```

lint results : 

```css
.c-slider__nav .c-btn {} 
/* warn 'a component cannot host another component' */
.c-popup.is-open {} 
/* good */
.c-popup .anyothernamespace-anyblockname {} 
/* warn 'this element should have a registered type/namespace' */
/* warn 'a component cannot host  anyothernamespace' */
.x-block {} 
/* warn 'this element should have a registered type/namespace' */
._hack .someExternalComponent {} 
/* good */
```

there is obviously many uses case that are not taken in account here, and the naming conventions that I used is certainly not the best, but you get the spirit…

Please let me know if what I want already exists or can be achieved with a combination of plugins, if any one want to help me create anything of that sort, or if there is a better place to go to make this a reality. 

cheers :)

GB

**Comment by jeddy3** [comment]
@gael-boyenval Thanks for writing up this request. It looks like you've a great outline for [a plugin](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/plugins.md). Supporting a particular methology was one of the reasons the plugin system was built:

> Plugins are rules and sets of rules built by the community that support **methodologies**, toolsets, non-standard CSS features, or very specific use cases.

I don't believe a plugin for BEMIT exists yet. At least no one has submitted one to [the list of plugins](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/plugins.md). And I don't believe such a metholodogy is enforcable with existing rules/plugins.

> if any one want to help me create anything of that sort,

Feel free to ask questions on Stackoverflow. Looking at [`selector-bem-pattern`](https://github.com/davidtheclark/stylelint-selector-bem-pattern) would be a good place to start. Perhaps there's scope to build on top of what's already there.

Labelling as plugin and closing. If you do create a new plugin for this specific methodolgy, please drop in a PR adding it to the list.

Good luck!

**Comment by gael-boyenval** [comment]
@jeddy3 : thank's for your guidance !
